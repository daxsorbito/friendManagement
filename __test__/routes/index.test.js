import request from 'supertest';
import { server, mongoose } from '../../index';

describe('routes test', () => {
    afterAll(async() => {
        await mongoose.disconnect();
    });
    afterEach(async() => {
        await mongoose.models.Friends.remove();
    });

    describe('POST# /friends', () => {
        it('should pass if correct data is passed', async() => {
            const result = await request(server)
                .post('/friends').send({ friends: ['test1@test.com', 'test2@test.com'] });
            expect(result.body).toMatchSnapshot();
        });

        it('should fail if content type is not json', async() => {
            const result = await request(server).post('/friends').send('test');
            expect(result.statusCode).toEqual(400);
            expect(result.body).toMatchSnapshot();
        });

        it('should fail if only one email is supplied in the payload', async() => {
            const result = await request(server).post('/friends')
                .send({ friends: ['test1@test.com'] });
            expect(result.statusCode).toEqual(400);
            expect(result.body).toMatchSnapshot();
        });

        it('should fail if payload is not array', async() => {
            const result = await request(server).post('/friends')
                .send({ friends: { email: 'test1@test.com' } });
            expect(result.statusCode).toEqual(400);
            expect(result.body).toMatchSnapshot();
        });
    });

    describe('GET# /friends/:email', () => {
        it('should pass if email is found', async() => {
            await request(server)
                .post('/friends').send({ friends: ['test1@test.com', 'test2@test.com'] });

            const result = await request(server).get('/friends/test1@test.com');

            expect(result.body).toMatchSnapshot();
        });

        it('should fail if email is invalid', async() => {
            const result = await request(server).get('/friends/INVALID_EMAIL');

            expect(result.statusCode).toEqual(400);
            expect(result.body).toMatchSnapshot();
        });

        it('should fail if email is not found', async() => {
            const result = await request(server).get('/friends/test1@test.com');

            expect(result.statusCode).toEqual(404);
            expect(result.body).toMatchSnapshot();
        });
    });

    describe('POST# /friends/common', () => {
        it('should pass if both has common friends', async() => {
            await request(server)
                .post('/friends').send({ friends: ['test1@test.com', 'test2@test.com', 'test3@test.com'] });

            const result = await request(server)
                .post('/friends/common')
                .send({ friends: ['test1@test.com', 'test3@test.com'] });

            expect(result.body).toMatchSnapshot();
        });

        it('should pass if there are no common friends', async() => {
            await request(server)
                .post('/friends').send({ friends: ['test1@test.com', 'test2@test.com', 'test3@test.com'] });

            const result = await request(server)
                .post('/friends/common')
                .send({ friends: ['test1@test.com', 'test5@test.com'] });

            expect(result.body).toMatchSnapshot();
        });
    });

    describe('POST# /friends/subscribe', () => {
        it('should pass if both requestor and target are valid', async() => {
            await request(server)
                .post('/friends').send({ friends: ['test1@test.com', 'test2@test.com', 'test3@test.com'] });

            const result = await request(server)
                .post('/friends/subscribe')
                .send({ requestor: 'test2@test.com', target: 'test1@test.com' });

            expect(result.body).toMatchSnapshot();
        });

        it('should not pass if requestor is not found', async() => {
            await request(server)
                .post('/friends').send({ friends: ['test1@test.com', 'test2@test.com', 'test3@test.com'] });

            const result = await request(server)
                .post('/friends/subscribe')
                .send({ requestor: 'NOTFOUND@test.com', target: 'test1@test.com' });

            expect(result.body).toMatchSnapshot();
        });

        it('should not pass if target is not found', async() => {
            await request(server)
                .post('/friends').send({ friends: ['test1@test.com', 'test2@test.com', 'test3@test.com'] });

            const result = await request(server)
                .post('/friends/subscribe')
                .send({ requestor: 'test1@test.com', target: 'NOTFOUND@test.com' });

            expect(result.body).toMatchSnapshot();
        });
    });

    describe('POST# /friends/block', () => {
        it('should pass if both requestor and target are valid', async() => {
            await request(server)
                .post('/friends').send({ friends: ['test1@test.com', 'test2@test.com', 'test3@test.com'] });

            const result = await request(server)
                .post('/friends/block')
                .send({ requestor: 'test2@test.com', target: 'test1@test.com' });

            expect(result.body).toMatchSnapshot();
        });

        it('should not pass if requestor is not found', async() => {
            await request(server)
                .post('/friends').send({ friends: ['test1@test.com', 'test2@test.com', 'test3@test.com'] });

            const result = await request(server)
                .post('/friends/block')
                .send({ requestor: 'NOTFOUND@test.com', target: 'test1@test.com' });

            expect(result.body).toMatchSnapshot();
        });

        it('should not pass if target is not found', async() => {
            await request(server)
                .post('/friends').send({ friends: ['test1@test.com', 'test2@test.com', 'test3@test.com'] });

            const result = await request(server)
                .post('/friends/block')
                .send({ requestor: 'test1@test.com', target: 'NOTFOUND@test.com' });

            expect(result.body).toMatchSnapshot();
        });
    });


});