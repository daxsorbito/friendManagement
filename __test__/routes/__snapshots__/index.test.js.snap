// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routes test GET# /friends/:email should fail if email is invalid 1`] = `
Object {
  "code": "BadRequest",
  "message": "Unexpected error.; caused by Error: Invalid email address.",
}
`;

exports[`routes test GET# /friends/:email should fail if email is not found 1`] = `
Object {
  "code": "ResourceNotFound",
  "message": "Resource not found.",
}
`;

exports[`routes test GET# /friends/:email should pass if email is found 1`] = `
Object {
  "count": 1,
  "friends": Array [
    "test2@test.com",
  ],
  "success": true,
}
`;

exports[`routes test POST# /friends should fail if content type is not json 1`] = `
Object {
  "code": "InvalidContent",
  "message": "Expects 'application/json'",
}
`;

exports[`routes test POST# /friends should fail if only one email is supplied in the payload 1`] = `
Object {
  "code": "BadRequest",
  "message": "Unexpected error.; caused by Error: Insufficient number of email addresses supplied.",
}
`;

exports[`routes test POST# /friends should fail if payload is not array 1`] = `
Object {
  "code": "BadRequest",
  "message": "Unexpected error.; caused by Error: Invalid friends payload.",
}
`;

exports[`routes test POST# /friends should pass if correct data is passed 1`] = `
Object {
  "success": true,
}
`;

exports[`routes test POST# /friends/block should not pass if requestor is not found 1`] = `
Object {
  "code": "ResourceNotFound",
  "message": "Resource not found. --requestor",
}
`;

exports[`routes test POST# /friends/block should not pass if target is not found 1`] = `
Object {
  "code": "ResourceNotFound",
  "message": "Resource not found. --target",
}
`;

exports[`routes test POST# /friends/block should pass if both requestor and target are valid 1`] = `
Object {
  "success": true,
}
`;

exports[`routes test POST# /friends/common should pass if both has common friends 1`] = `
Object {
  "count": 1,
  "friends": Array [
    "test2@test.com",
  ],
  "success": true,
}
`;

exports[`routes test POST# /friends/common should pass if there are no common friends 1`] = `
Object {
  "count": 0,
  "friends": Array [],
  "success": true,
}
`;

exports[`routes test POST# /friends/subscribe should not pass if requestor is not found 1`] = `
Object {
  "code": "ResourceNotFound",
  "message": "Resource not found. --requestor",
}
`;

exports[`routes test POST# /friends/subscribe should not pass if target is not found 1`] = `
Object {
  "code": "ResourceNotFound",
  "message": "Resource not found. --target",
}
`;

exports[`routes test POST# /friends/subscribe should pass if both requestor and target are valid 1`] = `
Object {
  "success": true,
}
`;
